### ISSUE TITLE
any idea to re-use the same error variable?

### ISSUE THREAD

**Issue by opensas** [arthurfiorette/proposal-safe-assignment-operator#11]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Versions

n/a

### A minimal reproducible example

n/a

### Description

Can you think of some way to use the same error variable, something like this (I know it doesn't work like this, it's just an example)

```javascript
async function getData() {
  const [err, response] ?= await fetch(
    "https://api.example.com/data"
  )

  if (err) {
    handleRequestError(err)
    return
  }

  const [err, json] ?= await response.json()

  if (err) {
    handleParseError(err)
    return
  }
  [...]
}
```

### Steps to Reproduce

n/a

**Comment by JAForbes** [comment]
@pygy taught me this, but JS already supports updating a previous destructured binding

```js
let [a,b] = [1,2]

console.log(a,b)
// logs 1,2

void ([a,b] = [4,5])

console.log(a,b)
// logs 4,5
```

So you'd just do the same thing:

```js
async function getData() {
  let [err, response] ?= await fetch(
    "https://api.example.com/data"
  )

  if (err) {
    handleRequestError(err)
    return
  }

  void ([err, json] ?= await response.json())

  if (err) {
    handleParseError(err)
    return
  }
  [...]
}
```

**Comment by BThero** [comment]
@JAForbes is it necessary to put ‘void (…)’ around the second assignment?

**Comment by JAForbes** [comment]
@BThero nope, if you use semi-colons carefully you can skip the parens and the void

```js
let [a,b] = [1,2];

console.log(a,b);
// logs 1,2

[a,b] = [4,5];

console.log(a,b);
// logs 4,5
```

Just be careful, starting a new line with `[...]` without a semi colon on the previous expression will be treated as a property access, and starting a new line with `(...)` will be interpreted as a function call.

**Comment by smeijer** [comment]
Back to `var` 😅

```js
async function getData() {
  var [err, response] ?= await fetch('https://api.example.com/data');
  if (err) return handleRequestError(err);

  var [err, json] ?= await response.json();
  if (err) return handleParseError(err);

  // ...
}
```

**Comment by BThero** [comment]
@JAForbes Ahh that's great. Yet another reason to enforce semicolons 😄

Thanks for the detailed explanation!

**Comment by arthurfiorette** [comment]
Maybe this is a valid reason to use the third approach `try` (as `using`), where the try operator would allow the overload of the error variable.

```ts
try [error, response] = await fetch('url')

if(err) { return 'error' }

try [error, json] = await response.json()
```

**Comment by arthurfiorette** [comment]
Reusing error variables is super confusing. I don't ever see this becoming a thing.

**Comment by felippe-regazio** [comment]
It doesn't fit all the scenarios, but you can use labeled statements in some cases:

```js
fizz: {
  const [ error, result ] ?= fnA();
  ...
}

buzz: {
  const [ error, result ] ?= fnB();
  ...
}
```

**Comment by JAForbes** [comment]
> It doesn't fit all the scenarios, but you can use labeled statements in some cases:
> 
> ```js
> fizz: {
>   const [ error, result ] ?= fnA();
>   ...
> }
> 
> buzz: {
>   const [ error, result ] ?= fnB();
>   ...
> }
> ```
> 
> But I can't see a practical reason for this since that we can just vary the variable name in the Iterable.

True, or even just unlabeled blocks:

```js
{
  const [ error, result ] ?= fnA();
  ...
}

{
  const [ error, result ] ?= fnB();
  ...
}
```

**Comment by khaosdoctor** [comment]
> > It doesn't fit all the scenarios, but you can use labeled statements in some cases:
> > ```js
> > fizz: {
> >   const [ error, result ] ?= fnA();
> >   ...
> > }
> > 
> > buzz: {
> >   const [ error, result ] ?= fnB();
> >   ...
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But I can't see a practical reason for this since that we can just vary the variable name in the Iterable.
> 
> True, or even just unlabeled blocks:
> 
> ```js
> {
>   const [ error, result ] ?= fnA();
>   ...
> }
> 
> {
>   const [ error, result ] ?= fnB();
>   ...
> }
> ```

I would prefer to use `try` as using in this case over separated scopes, aesthetically speaking. But I think (and as it was stated on #4) this implies a different handling of the variable which could be a bigger scope, so I'd go with parallel assignments in this one

**Comment by arthurfiorette** [comment]
Since the new rewriting, this is possible.

https://github.com/arthurfiorette/proposal-safe-assignment-operator/pull/53

Feel free to reopen it if I missed something :)
